lib/deploy/stepFunctions/compileIamRole.js

```diff
@@ -245,25 +245,82 @@ function getBatchDynamoDBPermissions(action, state) {
+      }
+    } else if (_.has(state, ['Parameters', 'WorkgroupName.$'])) {
+      workgroupArn = { 'Fn::Sub': 'arn:${AWS::Partition}:redshift-serverless:${AWS::Region}:${AWS::AccountId}:workgroup/*' };
+    } else if (_.has(state, ['Parameters', 'ClusterIdentifier'])) {
+      clusterName = state.Parameters.ClusterIdentifier;
+    } else {
+      clusterName = '*';
+    }
+
+    let secretArn;
+    let dbUser;
+    if (_.has(state, ['Parameters', 'SecretArn'])) {
+      if (state.Parameters.SecretArn.startsWith('arn:')) {
+        secretArn = state.Parameters.SecretArn;
+      } else {
+        secretArn = { 'Fn::Sub': `arn:\${AWS::Partition}:secretsmanager:\${AWS::Region}:\${AWS::AccountId}:secret:${state.Parameters.SecretArn}*` };
+      }
+    } else if (_.has(state, ['Parameters', 'SecretArn.$'])) {
+      secretArn = { 'Fn::Sub': 'arn:${AWS::Partition}:secretsmanager:${AWS::Region}:${AWS::AccountId}:secret:*' };
+    } else if (_.has(state, ['Parameters', 'DbUser'])) {
+      dbUser = state.Parameters.DbUser;
+    } else if (_.has(state, ['Parameters', 'DbUser.$'])) {
+      dbUser = '*';
+    }
+
+    permissions.push({
       action,
-      resource: { 'Fn::Sub': `arn:\${AWS::Partition}:redshift:\${AWS::Region}:\${AWS::AccountId}:cluster:${clusterName}` },
-    }, {
-      action: 'redshift:GetClusterCredentials',
-      resource: [
-        { 'Fn::Sub': `arn:\${AWS::Partition}:redshift:\${AWS::Region}:\${AWS::AccountId}:dbname:${clusterName}/${dbName}` },
-        { 'Fn::Sub': `arn:\${AWS::Partition}:redshift:\${AWS::Region}:\${AWS::AccountId}:dbuser:${clusterName}/${dbUser}` },
-      ],
-    }];
+      resource: workgroupArn || { 'Fn::Sub': `arn:\${AWS::Partition}:redshift:\${AWS::Region}:\${AWS::AccountId}:cluster:${clusterName}` },
+    });
+
+    if (secretArn) {
+      permissions.push({
+        action: 'secretsmanager:GetSecretValue',
+        resource: secretArn,
+      });
+    } else if (dbUser) {
+      permissions.push({
+        action: 'redshift:GetClusterCredentials',
+        resource: [
+          { 'Fn::Sub': `arn:\${AWS::Partition}:redshift:\${AWS::Region}:\${AWS::AccountId}:dbuser:${clusterName}/${dbUser}` },
+          { 'Fn::Sub': `arn:\${AWS::Partition}:redshift:\${AWS::Region}:\${AWS::AccountId}:dbname:${clusterName}/${dbName}` },
+        ],
+      });
+    } else {
+      if (workgroupArn) { // eslint-disable-line no-lonely-if
+        permissions.push({
+          action: 'redshift-serverless:GetCredentials',
+          resource: workgroupArn,
+        });
+      } else {
+        permissions.push({
+          action: 'redshift:GetClusterCredentialsWithIAM',
+          resource: { 'Fn::Sub': `arn:\${AWS::Partition}:redshift:\${AWS::Region}:\${AWS::AccountId}:dbname:${clusterName}/${dbName}` },
+        });
+      }
+    }
+  } else {
+    permissions.push({
+      action,
+      resource: '*',
```

Comment:
How do you reach this section under what conditions? Does this mean this section add * permission to the generated IAM by this plugin?

Reply:
@horike37 Thank you for the review.

This statement is executed when the normalized Resource field value of a Task state is one of the following ARNs.

- `arn:aws:states:::aws-sdk:redshiftdata:listStatements`
- `arn:aws:states:::aws-sdk:redshiftdata:describeStatement`
- `arn:aws:states:::aws-sdk:redshiftdata:getStatementResult`
- `arn:aws:states:::aws-sdk:redshiftdata:cancelStatement`

The corresponding actions do not support resource-level permissions, therefore requiring the wildcard `*` to be specified as the resource in the policy ([documentation](https://docs.aws.amazon.com/service-authorization/latest/reference/list_amazonredshiftdataapi.html#amazonredshiftdataapi-actions-as-permissions)). The resulting statement would for example look like this.

```
{
  "Effect": "Allow",
  "Action": [
    "redshift-data:ListStatements"
  ],
  "Resource": "*"
}
```

The `getRedshiftDataPermissions` function, which includes the code in question, is called by the `getIamPermissions` function from within lines 607-618, which are not changed in this pull request.

Reply:
Thank you for the explanation! I got it.